File: sublimate/view/window.py
# -*- coding: utf-8 -*-
from sublimate.toolkit import Widget, OverlayMixin

from .menubar import Menubar
from .sidebar import Sidebar
from .tabgroup import TabGroup
from .statusbar import Statusbar


class Window(Widget, OverlayMixin):

    def __init__(self, app):
        self.menubar = self.create_widget(Menubar, app.get_menu('Main'))
        self.sidebar = self.create_widget(Sidebar, app.project)
        self.tabgroup = self.create_widget(TabGroup, app.group)
        self.statusbar = self.create_widget(Statusbar)

    # def on_mouse(self, event): # todo: not just on_mouse
        # self.editor.capture_focus()

    def render(self, canvas):
        menubar_canvas, frame_canvas, statusbar_canvas = \
            canvas.vert[self.menubar.height, ..., self.statusbar.height]
        sidebar_canvas, tabgroup_canvas = \
            frame_canvas.horz[25, ...]

        self.menubar.render(menubar_canvas)
        self.sidebar.render(sidebar_canvas)
        self.tabgroup.render(tabgroup_canvas)
        self.statusbar.render(statusbar_canvas)
        self.render_modals(canvas)

File: src/view/window.rs
use toolkit::*;
use core::Core;

use core::bindings::Key;

use view::menubar::Menubar;
use view::editor::Editor;
use view::sidebar::Sidebar;

#[derive(Debug)]
pub struct Window {
    core: Core,
    menubar: Menubar,
    editor: Editor,
    sidebar: Sidebar,
}

impl Window {
    pub fn new(core: Core) -> Window {
        let menubar = Menubar::new(&core);
        let sidebar = Sidebar::new(&core);
        let editor = Editor::new(&core);
        Window {
            core: core,
            menubar: menubar,
            sidebar: sidebar,
            editor: editor
        }
    }

    pub fn on_keypress(&mut self, mut canvas: Canvas, key: Key) {
        self.menubar.on_keypress(&self.core, canvas.cut_top(1), key);
    }


    pub fn render(&self, mut canvas: Canvas) {
        self.menubar.view(&self.core).render(canvas.cut_top(1));
        self.sidebar.view(&self.core).render(canvas.cut_left(30));
        self.editor.view(&self.core).render(canvas);
    }
}
